docs/adr/0041-gcp-cloud-tasks-for-job-queuing.md

```diff
@@ -0,0 +1,80 @@
+informed: FxA Team
+---
+
+# Use Google Cloud Tasks for job queueing
+
+## Context and Problem Statement
+
+Many account related tasks require interaction with multiple systems that may exceed the typical request deadline. Some of these tasks require additional guarantee's they fail, or may partially complete before erroring and require a retry. We also occasionally run into cases where there's many thousands or millions of tasks to complete that have been handled before by splitting up lists of arguments and running them in parallel across many nodes manually by our SRE's. Another challenge we've faced when running large quantities of tasks in a batch manner is ensuring rate limits or adverse system conditions aren't hit.
+
+This type of task handling is typically handled in large systems by a job/task/queue system, which can track jobs, retry them if they fail, inspect failed jobs, and limit the rate at which they're processed. FxA has used SQS/SNS for messaging for some task/event distribution, but has no system built for queueing tasks and processing jobs in a reliable manner.
+
+The most immediate problem is that we wants to run large quantities of tasks in a reliable manner, with effective rate limiting. Later, we'd like to be able to queue tasks such as sending an email so that they complete reliably and do not hold up request processing, ideally in a manner that is not programming language dependent so that emails can be sent separately by the Sub Plat team without needing to replicate email handling functionality.
+
+We have had problems in the past using message queues as neither Google Pub/Sub nor AWS SQS have a way to inspect messages that may be stuck in the queue.
+
+## Decision Drivers
+
+- Rate limiting of job processing
+- Job retries
+- Avoid additional services requiring operational support
+- Task deduplication
+- Task inspection while in queue
+- Scheduling of tasks in the future
+- Programming language agnostic
+
+## Considered Options
+
+Push based system can push the tasks in the queue out to the task handler that processes it, which avoids the need to setup a separate service of task processors that would pull from a the job queue. Pull based will require a separate deployment by SRE of job processors to pull tasks from the queue.
+
+- Google Cloud Tasks (Push based)
+- Google Pub/Sub (Push based)
+- BullMQ + Redis (Pull based)
+- Database table of tasks (Pull based)
+
+## Decision Outcome
+
+Chosen option: Google Cloud Tasks, because it meets all our criteria in a cost effective manner with minimal risk and we already utilize a variety of Google Cloud Platform services.
+
+## Pros and Cons of the Options
+
+### Google Cloud Tasks
+
+- Good, because we can have our existing API nodes process tasks with no additional job processor service.
+- Good, because centralized rate limiting means we can easily rate limit the overall processing of tasks regardless of how many nodes are processing them.
```

Comment:
is the rate limiting controlled in the code or on a dashboard somewhere in google console?

Reply:
something you (well, whoever has access) can edit in the GCP console.